content/ros/ROS_Nodes_Topics_Messages/presentation/ros_nodes_topics_messgaes.tex

```diff
@@ -0,0 +1,1560 @@
+
+\begin{frame}[fragile]{Writing a subscriber node in Python}
+    \framesubtitle{../scripts/03\_simple\_subscriber.py}
+    \lstset{language=python,
+        basicstyle=\scriptsize,
+        keywordstyle=\color{blue}\ttfamily,
+        stringstyle=\color{red}\ttfamily,
+        commentstyle=\color{green}\ttfamily,
+        morecomment=[l][\color{magenta}]{\#},
+        showstringspaces=false
+    }
+    \begin{lstlisting}
+    
+    #!/usr/bin/env python
+    
+    import rospy
+    from std_msgs.msg import String
+
+    def my_callback_function(msg):
+        print msg
+    
+    rospy.init_node('listener')
+    rate = rospy.Rate(100) 
+    sub = rospy.Subscriber('myFirstTopic', String,
+                            callback=my_callback_function)
+    
+    while not rospy.is_shutdown():
+        pass
+    
+    
+    \end{lstlisting}
+\end{frame}
+
+
+\begin{frame}[fragile]{Writing a subscriber node in Python}
+        \framesubtitle{../scripts/04\_simple\_subscriber.py}
+        
+    \lstset{language=python,
+        basicstyle=\scriptsize,
+        keywordstyle=\color{blue}\ttfamily,
+        stringstyle=\color{red}\ttfamily,
+        commentstyle=\color{green}\ttfamily,
+        morecomment=[l][\color{magenta}]{\#},
+        showstringspaces=false
+    }
+    \begin{lstlisting}
+#!/usr/bin/env python
+
+import rospy
+from std_msgs.msg import String
+
+
+def my_callback_function(msg):
+    print msg
+
+rospy.init_node('listener')
+rate = rospy.Rate(100)
+
+rospy.Subscriber('myFirstTopic', String, callback=my_callback_function)
+
+rospy.spin()
+
+    
+    \end{lstlisting}
+\end{frame}
+
+
+
+\begin{frame}[fragile]{Writing a subscriber node in Python}
+    \framesubtitle{Subscriber class}
+    
+    
+    \begin{itemize}
+        \centering
+        \item go to the README, and let's do section:\\ 
+        \textbf{Simple subscriber}
+    \end{itemize} 
+\end{frame}
+
+
+\begin{frame}[fragile]{Writing a subscriber node in Python}
+    \framesubtitle{Subscriber class}
+    \begin{focus}
+        
+        \ttfamily rospy.Subscriber(\\
+        {\color{red}name},\\
+        {\color{red}data\_class},\\
+        {\color{red}callback=None}, \\
+        {\color{blue}callback\_args=None},\\
+        {\color{blue}queue\_size=None},\\
+        {\color{blue}buff\_size=65536},\\
+        {\color{blue}tcp\_nodelay=False}\\)
+    \end{focus}
+    
+\end{frame}
+
+
+
+
+\begin{frame}[fragile]{Writing a subscriber node in Python}
+    \framesubtitle{Subscriber class}
+    
+    
+    \begin{itemize}
+        \centering
+        \item go to the README, and let's do section:\\ 
+        \textbf{Simple subscriber}
+    \end{itemize} 
+\end{frame}
+
+
+\subsection{General notes}
+\begin{frame}{General notes}
+    \framesubtitle{more on ROS nodes}
+    \begin{itemize}
+        \item You can define multiple publishers and subscribers in a node.
+        \item You can call {\ttfamily \colorbox{gray!30!white}{init\_node}} function once only!
```

Comment:
Unless you use background threads.

Reply:
Done (I added your comment). 

Do you mean if you have multiple threads, you can call it multiple times?

Reply:
More or less. For instance, you can start the node on a background thread, do some monitoring (e.g. if the ROS master is alive) and, if necessary, restart the node again (e.g. if the master was dead and comes back).

In case you're interested, we have an example of this here: https://github.com/ropod-project/black-box/blob/master/pybb/black_box/datalogger/data_readers/rostopic_reader.py#L56 (we use background processes instead of background threads there; there were other reasons for this design decision).

Reply:
Ok I got it. I modified the slides again mentioning:

> If you are spawning multiple processes in your script, then you can call it once for each process).